### PR TITLE
Upgrade to TensorFlow 0.12.0

### PR DESCRIPTION
**Pull request by adamcrume**

### CODE REVIEW

**Comment by adamcrume** [comment]
Fixes https://github.com/tensorflow/rust/issues/31, and then some.

File: travis-ci/install.sh [hunk 1]
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-BAZEL_URL=https://github.com/bazelbuild/bazel/releases/download/0.3.0/bazel-0.3.0-installer-linux-x86_64.sh
+BAZEL_URL=https://github.com/bazelbuild/bazel/releases/download/0.3.2/bazel-0.3.2-installer-linux-x86_64.sh
```
Comment (jhseu) [inline travis-ci/install.sh]: Use the 0.4.3 release? I've personally run into strange bugs on the 0.3.x release.

File: src/graph.rs [hunk 2]
```diff
@@ -126,6 +151,71 @@ impl Graph {
...
+
+  /// Returns the shape of the Tensor referenced by `output`.
+  ///
+  /// If the number of dimensions in the shape is unknown, the return value is
+  /// None. Otherwise, each element of `dims` will be set corresponding to the
+  /// size of the dimension. An unknown dimension is represented by `-1`.
+  ///
+  /// Returns an error if:
+  ///   * `output` is not in `graph`.
+  pub fn tensor_shape(&self, output: Output) -> Result<Option<Vec<i64>>> {
```
Comment (jhseu) [inline src/graph.rs]: I'd favor returning TensorShape type where fetching a dimension is Option<i64> and is None if the shape of that dim is unknown.

Comment (adamcrume) [inline src/graph.rs]: So something roughly like `type TensorShape = Vec<Option<i64>>` and have `tensor_shape()` return `Result<Option<TensorShape>>`?

**Comment by jhseu** [comment]
`Result<TensorShape>` because a shape with an unknown number of dimensions is still valid.

TensorShape would have two functions (naming up to you):
`dim(i) -> Option<int64>`
`num_dims() -> Option<int64>`

File: travis-ci/install.sh (hunk 1 above)
Comment (adamcrume) [inline travis-ci/install.sh]: Easy enough. :)

**Comment by adamcrume** [comment]
I added a `TensorShape` type which is the parameter type for `OperationDescription::set_attr_shape()` and the return type for `Graph::tensor_shape()`

File: src/lib.rs [hunk 3]
```diff
@@ -1022,14 +1021,78 @@ trait OperationTrait {
...
+
+/// A TensorShape is the shape of a tensor.  A TensorShape may be an unknown
+/// rank, or it may have a known rank with each dimension being known or
+/// unknown.
+#[derive(Debug,Eq,Ord,PartialEq,PartialOrd,Hash,Clone)]
+pub struct TensorShape(Option<Vec<Option<i64>>>);
+
+impl TensorShape {
+  /// Returns the number of dimensions if known, or None if unknown.
+  pub fn rank(&self) -> Option<usize> {
```
Comment (jhseu) [inline src/lib.rs]: We're generally discouraging the use of the term `rank`:
https://www.tensorflow.org/how_tos/documentation/
`Refer to tensors with 3 or more dimensions as 3-D tensors or n-D tensors. Use the word "rank" only if it makes sense, but try to use "dimension" instead. Never use the word "order" to describe the size of a tensor.`

In the C++ class, we're using the convention dims() and dim_size().
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/tensor_shape.h

Comment (adamcrume) [inline src/lib.rs]: Renamed to dims().  Out of curiosity, do you know why "rank" is discouraged?  Is it just to keep the jargon to a minimum?

**Pull request merged**